In [1]:
!pip install psycopg2


[notice] A new release of pip is available: 26.0.1 -> 26.1.2
[notice] To update, run: python.exe -m pip install --upgrade pip


In [2]:
import psycopg2
import pandas as pd

In [ ]:
DBNAME = 'your_dbname'
USER = 'your_username'
PASSWORD = 'your_password'
HOST = 'your_host'
PORT = 5432

connection = psycopg2.connect(
   dbname=DBNAME,
   user=USER,
   host=HOST,
   password=PASSWORD,
   port=PORT
)

In [5]:
# Задание 3.1: Сколько вакансий есть в базе?
query = '''
SELECT COUNT(*) AS vacancies_count
FROM public.vacancies
'''
df_3_1 = pd.read_sql_query(query, connection)
df_3_1

C:\Users\Lenovo\AppData\Local\Temp\ipykernel_20144\1946259452.py:6: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_3_1 = pd.read_sql_query(query, connection)


,vacancies_count
0,49197


In [6]:
# Задание 3.2: Сколько работодателей?
query = '''
SELECT COUNT(*) AS employers_count
FROM public.employers
'''
df_3_2 = pd.read_sql_query(query, connection)
df_3_2

C:\Users\Lenovo\AppData\Local\Temp\ipykernel_20144\3202135418.py:6: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_3_2 = pd.read_sql_query(query, connection)


,employers_count
0,23501


In [7]:
# Задание 3.3: Сколько регионов?
query = '''
SELECT COUNT(*) AS areas_count
FROM public.areas
'''
df_3_3 = pd.read_sql_query(query, connection)
df_3_3

C:\Users\Lenovo\AppData\Local\Temp\ipykernel_20144\2560798308.py:6: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_3_3 = pd.read_sql_query(query, connection)


,areas_count
0,1362


In [8]:
# Задание 3.4: Сколько сфер деятельности?
query = '''
SELECT COUNT(*) AS industries_count
FROM public.industries
'''
df_3_4 = pd.read_sql_query(query, connection)
df_3_4

C:\Users\Lenovo\AppData\Local\Temp\ipykernel_20144\597811072.py:6: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_3_4 = pd.read_sql_query(query, connection)


,industries_count
0,294


В базе данных содержится 49 197 вакансий от 23 501 работодателя. География представлена 1 362 регионами, а сферы деятельности охватывают 294 категории. Такое соотношение (в среднем ~2 вакансии на работодателя) говорит о том, что подавляющее большинство компаний в базе — это небольшие или средние работодатели с единичными открытыми позициями, а не крупные корпорации с массовым набором.

In [9]:
# Задание 4.1: Количество вакансий в каждом регионе, топ-5
query = '''
SELECT 
    a.name AS area,
    COUNT(*) AS cnt
FROM public.vacancies v
JOIN public.areas a ON v.area_id = a.id
GROUP BY a.name
ORDER BY cnt DESC
LIMIT 5
'''
df_4_1 = pd.read_sql_query(query, connection)
df_4_1

C:\Users\Lenovo\AppData\Local\Temp\ipykernel_20144\3986725355.py:12: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_4_1 = pd.read_sql_query(query, connection)


,area,cnt
0,Москва,5333
1,Санкт-Петербург,2851
2,Минск,2112
3,Новосибирск,2006
4,Алматы,1892


In [10]:
# Задание 4.2: Количество вакансий с заполненным хотя бы одним полем зарплаты
query = '''
SELECT COUNT(*) AS cnt
FROM public.vacancies
WHERE salary_from IS NOT NULL 
   OR salary_to IS NOT NULL
'''
df_4_2 = pd.read_sql_query(query, connection)
df_4_2

C:\Users\Lenovo\AppData\Local\Temp\ipykernel_20144\4271808447.py:8: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_4_2 = pd.read_sql_query(query, connection)


,cnt
0,24073


In [11]:
# Задание 4.3: Средние значения нижней и верхней границы зарплатной вилки
query = '''
SELECT 
    ROUND(AVG(salary_from)::numeric) AS avg_salary_from,
    ROUND(AVG(salary_to)::numeric) AS avg_salary_to
FROM public.vacancies
'''
df_4_3 = pd.read_sql_query(query, connection)
df_4_3

C:\Users\Lenovo\AppData\Local\Temp\ipykernel_20144\3240757825.py:8: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_4_3 = pd.read_sql_query(query, connection)


,avg_salary_from,avg_salary_to
0,71065.0,110537.0


In [12]:
# Задание 4.4: Количество вакансий по сочетанию schedule + employment, топ-2
query = '''
SELECT 
    schedule,
    employment,
    COUNT(*) AS cnt
FROM public.vacancies
GROUP BY schedule, employment
ORDER BY cnt DESC
LIMIT 2
'''
df_4_4 = pd.read_sql_query(query, connection)
df_4_4

C:\Users\Lenovo\AppData\Local\Temp\ipykernel_20144\1596183341.py:12: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_4_4 = pd.read_sql_query(query, connection)


,schedule,employment,cnt
0,Полный день,Полная занятость,35367
1,Удаленная работа,Полная занятость,7802


In [13]:
# Задание 4.5: Требуемый опыт работы в порядке возрастания количества вакансий
query = '''
SELECT 
    experience,
    COUNT(*) AS cnt
FROM public.vacancies
GROUP BY experience
ORDER BY cnt
'''
df_4_5 = pd.read_sql_query(query, connection)
df_4_5

C:\Users\Lenovo\AppData\Local\Temp\ipykernel_20144\554414914.py:10: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_4_5 = pd.read_sql_query(query, connection)


,experience,cnt
0,Более 6 лет,1337
1,Нет опыта,7197
2,От 3 до 6 лет,14511
3,От 1 года до 3 лет,26152


Рынок вакансий сильно сконцентрирован географически: пятёрку лидеров формируют крупнейшие мегаполисы региона (Москва, Санкт-Петербург, Минск, Новосибирск, Алматы) — Москва почти вдвое опережает второе место, что ожидаемо для столицы с наибольшей концентрацией IT- и корпоративного сектора.
Информация о зарплате указана лишь примерно в половине вакансий (24 073 из 49 197) — работодатели нередко скрывают вилку, что стоит учитывать при любом зарплатном анализе как ограничение выборки. Средняя нижняя граница вилки — 71 065 ₽, верхняя — 110 537 ₽; разброс почти в 1.5 раза отражает типичную неопределённость в объявлениях о вакансиях.
Абсолютно доминирующий формат работы — «Полный день + Полная занятость» (35 367 вакансий, более 70% всей базы), удалённая полная занятость идёт вторым с большим отрывом (7 802) — это показывает, что удалёнка, при всей своей популярности в обсуждениях, всё ещё остаётся нишевым, а не основным форматом на рынке в целом.
По требуемому опыту наибольший спрос — на специалистов с опытом от 1 года до 3 лет (26 152 вакансии), а вакансий для senior-уровня (более 6 лет) — меньше всего (1337). Это говорит о том, что рынок труда преимущественно ориентирован на специалистов среднего уровня, а не на новичков без опыта или высококвалифицированных экспертов.

In [14]:
# Задание 5.1: Работодатели на первом и пятом месте по количеству вакансий
query = '''
SELECT 
    e.name AS employer,
    COUNT(*) AS cnt
FROM public.vacancies v
JOIN public.employers e ON v.employer_id = e.id
GROUP BY e.name
ORDER BY cnt DESC
LIMIT 5
'''
df_5_1 = pd.read_sql_query(query, connection)
df_5_1

C:\Users\Lenovo\AppData\Local\Temp\ipykernel_20144\3287820753.py:12: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_1 = pd.read_sql_query(query, connection)


,employer,cnt
0,Яндекс,1933
1,Ростелеком,491
2,Тинькофф,444
3,СБЕР,428
4,Газпром нефть,331


In [15]:
# Задание 5.2: Количество работодателей и вакансий по каждому региону;
# среди регионов без вакансий — тот, где больше всего работодателей
query = '''
WITH emp_counts AS (
    SELECT area AS area_id, COUNT(*) AS employers_cnt
    FROM public.employers
    GROUP BY area
),
vac_counts AS (
    SELECT area_id, COUNT(*) AS vacancies_cnt
    FROM public.vacancies
    GROUP BY area_id
)
SELECT 
    a.name AS area,
    COALESCE(ec.employers_cnt, 0) AS employers_cnt,
    COALESCE(vc.vacancies_cnt, 0) AS vacancies_cnt
FROM public.areas a
LEFT JOIN emp_counts ec ON ec.area_id = a.id
LEFT JOIN vac_counts vc ON vc.area_id = a.id
WHERE COALESCE(vc.vacancies_cnt, 0) = 0
ORDER BY employers_cnt DESC
LIMIT 1
'''
df_5_2 = pd.read_sql_query(query, connection)
df_5_2

C:\Users\Lenovo\AppData\Local\Temp\ipykernel_20144\87413827.py:25: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_2 = pd.read_sql_query(query, connection)


,area,employers_cnt,vacancies_cnt
0,Россия,410,0


In [16]:
# Задание 5.3: Количество уникальных регионов на работодателя, максимум
query = '''
SELECT 
    e.name AS employer,
    COUNT(DISTINCT v.area_id) AS regions_cnt
FROM public.vacancies v
JOIN public.employers e ON v.employer_id = e.id
GROUP BY e.name
ORDER BY regions_cnt DESC
LIMIT 1
'''
df_5_3 = pd.read_sql_query(query, connection)
df_5_3

C:\Users\Lenovo\AppData\Local\Temp\ipykernel_20144\871126622.py:12: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_3 = pd.read_sql_query(query, connection)


,employer,regions_cnt
0,Яндекс,181


In [17]:
# Задание 5.4: Количество работодателей без указанной сферы деятельности
query = '''
SELECT COUNT(*) AS cnt
FROM public.employers e
LEFT JOIN public.employers_industries ei ON e.id = ei.employer_id
WHERE ei.employer_id IS NULL
'''
df_5_4 = pd.read_sql_query(query, connection)
df_5_4

C:\Users\Lenovo\AppData\Local\Temp\ipykernel_20144\3916570737.py:8: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_4 = pd.read_sql_query(query, connection)


,cnt
0,8419


In [18]:
# Задание 5.5: Третья по алфавиту компания с ровно 4 сферами деятельности
query = '''
SELECT 
    e.name AS employer
FROM public.employers e
JOIN public.employers_industries ei ON e.id = ei.employer_id
GROUP BY e.name
HAVING COUNT(*) = 4
ORDER BY e.name
LIMIT 1 OFFSET 2
'''
df_5_5 = pd.read_sql_query(query, connection)
df_5_5

C:\Users\Lenovo\AppData\Local\Temp\ipykernel_20144\2675919834.py:12: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_5 = pd.read_sql_query(query, connection)


,employer
0,2ГИС


In [19]:
# Задание 5.6: Количество работодателей со сферой "Разработка программного обеспечения"
query = '''
SELECT COUNT(*) AS cnt
FROM public.employers_industries ei
JOIN public.industries i ON ei.industry_id = i.id
WHERE i.name = 'Разработка программного обеспечения'
'''
df_5_6 = pd.read_sql_query(query, connection)
df_5_6

C:\Users\Lenovo\AppData\Local\Temp\ipykernel_20144\1151961816.py:8: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_6 = pd.read_sql_query(query, connection)


,cnt
0,3553


In [20]:
# Задание 5.7: Вакансии Яндекса по регионам-миллионникам + строка Total
query = '''
WITH million_cities AS (
    SELECT unnest(ARRAY[
        'Москва','Санкт-Петербург','Новосибирск','Екатеринбург','Казань',
        'Нижний Новгород','Челябинск','Красноярск','Самара','Уфа',
        'Ростов-на-Дону','Омск','Краснодар','Воронеж','Пермь','Волгоград'
    ]) AS name
)
SELECT a.name, COUNT(*) AS cnt
FROM public.vacancies v
JOIN public.employers e ON v.employer_id = e.id
JOIN public.areas a ON v.area_id = a.id
WHERE e.name = 'Яндекс' AND a.name IN (SELECT name FROM million_cities)
GROUP BY a.name

UNION ALL

SELECT 'Total', COUNT(*)
FROM public.vacancies v
JOIN public.employers e ON v.employer_id = e.id
JOIN public.areas a ON v.area_id = a.id
WHERE e.name = 'Яндекс' AND a.name IN (SELECT name FROM million_cities)

ORDER BY cnt
'''
df_5_7 = pd.read_sql_query(query, connection)
df_5_7

C:\Users\Lenovo\AppData\Local\Temp\ipykernel_20144\2577350279.py:27: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_7 = pd.read_sql_query(query, connection)


,name,cnt
0,Омск,21
1,Челябинск,22
2,Красноярск,23
3,Волгоград,24
4,Пермь,25
5,Казань,25
6,Ростов-на-Дону,25
7,Самара,26
8,Уфа,26
9,Краснодар,30


Яндекс — безусловный лидер рынка вакансий: 1933 вакансии (в 4 раза больше второго места — Ростелекома с 491), присутствие в 181 регионе (тоже абсолютный максимум) и 485 вакансий только в городах-миллионниках. Это отражает масштаб компании как одного из крупнейших работодателей в IT-сфере региона с географически распределённой структурой найма.
Заметная часть базы работодателей (8419 из 23501, то есть почти 36%) вообще не указала сферу деятельности — это существенный пробел в данных, который стоит учитывать при любом отраслевом анализе: реальная картина по индустриям может отличаться от той, что видна по заполненным записям. Среди тех, кто сферу указал, «Разработка программного обеспечения» — одна из самых распространённых (3553 работодателя), что ожидаемо для базы, ориентированной на IT-вакансии.
Интересная находка: «Россия» как отдельная запись в справочнике регионов имеет 410 работодателей, но ни одной вакансии — вероятно, это работодатели, указавшие страну в целом без привязки к конкретному городу, что стоит иметь в виду при географическом анализе (часть работодателей "невидима" для регионального среза).

In [21]:
# Задание 6.1: Сколько вакансий имеет отношение к данным?
# Критерий: в названии есть 'data' или 'данн'
query = '''
SELECT COUNT(*) AS cnt
FROM public.vacancies
WHERE name ILIKE '%data%' OR name ILIKE '%данн%'
'''
df_6_1 = pd.read_sql_query(query, connection)
df_6_1

C:\Users\Lenovo\AppData\Local\Temp\ipykernel_20144\3022958183.py:8: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_1 = pd.read_sql_query(query, connection)


,cnt
0,1771


In [22]:
# Задание 6.2: Вакансии для начинающего дата-сайентиста (уровень Junior)
query = '''
SELECT COUNT(*) AS cnt
FROM public.vacancies
WHERE (
    name ILIKE '%data scientist%'
    OR name ILIKE '%data science%'
    OR name ILIKE '%исследователь данных%'
    OR name ~* '\\yml\\y'
    OR name ILIKE '%machine learning%'
    OR name ILIKE '%машинн%обучен%'
)
AND (
    name ILIKE '%junior%'
    OR experience = 'Нет опыта'
    OR employment = 'Стажировка'
)
'''
df_6_2 = pd.read_sql_query(query, connection)
df_6_2

C:\Users\Lenovo\AppData\Local\Temp\ipykernel_20144\2281104303.py:19: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_2 = pd.read_sql_query(query, connection)


,cnt
0,51


In [23]:
# Задание 6.6: Средняя зарплата для DS по каждому типу опыта (без Junior)
query = '''
SELECT 
    experience,
    ROUND(AVG(
        (COALESCE(salary_from, salary_to) + COALESCE(salary_to, salary_from)) / 2
    )::numeric) AS avg_salary
FROM public.vacancies
WHERE (
    name ILIKE '%data scientist%'
    OR name ILIKE '%data science%'
    OR name ILIKE '%исследователь данных%'
    OR name ~* '\\yml\\y'
    OR name ILIKE '%machine learning%'
    OR name ILIKE '%машинн%обучен%'
)
AND NOT (
    name ILIKE '%junior%'
    OR experience = 'Нет опыта'
    OR employment = 'Стажировка'
)
AND (salary_from IS NOT NULL OR salary_to IS NOT NULL)
GROUP BY experience
ORDER BY experience
'''
df_6_6 = pd.read_sql_query(query, connection)
df_6_6

C:\Users\Lenovo\AppData\Local\Temp\ipykernel_20144\3605346254.py:26: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_6 = pd.read_sql_query(query, connection)


,experience,avg_salary
0,От 1 года до 3 лет,150785.0
1,От 3 до 6 лет,243115.0


Вакансий, так или иначе связанных с данными, довольно много — 1771, но специализированных именно под дата-сайентиста заметно меньше: лишь 51 из них ориентирована на начинающего специалиста (Junior/без опыта/стажировка), что говорит о том, что рынок DS-вакансий преимущественно рассчитан на специалистов уже с каким-то опытом, а не на новичков — вход в профессию через прямые junior-вакансии DS ограничен.
Зарплатный разрыв между уровнями опыта показателен: специалист с опытом от 1 до 3 лет может рассчитывать в среднем на 150 785 ₽, а с опытом от 3 до 6 лет — уже на 243 115 ₽, то есть рост почти в 1.6 раза. Это подчёркивает, что в DS практический опыт ценится особенно высоко — разница ощутимо больше, чем обычно наблюдается при переходе между аналогичными грейдами в других IT-специальностях.
Обращает на себя внимание отсутствие данных по зарплате для сегмента "Более 6 лет" среди DS-вакансий — либо такие вакансии крайне редко указывают зарплатную вилку в принципе, либо среди отфильтрованных DS-вакансий записей с таким опытом почти не осталось после исключения Junior-сегмента.

In [24]:
# Задание 6.3: Количество DS-вакансий с ключевым навыком SQL или postgres
query = '''
SELECT COUNT(*) AS cnt
FROM public.vacancies v
WHERE (
    name ILIKE '%data scientist%'
    OR name ILIKE '%data science%'
    OR name ILIKE '%исследователь данных%'
    OR name ~* '\\yml\\y'
    OR name ILIKE '%machine learning%'
    OR name ILIKE '%машинн%обучен%'
)
AND NOT (
    name ILIKE '%junior%'
    OR experience = 'Нет опыта'
    OR employment = 'Стажировка'
)
AND (
    key_skills ILIKE '%SQL%'
    OR key_skills ILIKE '%postgres%'
)
'''
df_6_3 = pd.read_sql_query(query, connection)
df_6_3

C:\Users\Lenovo\AppData\Local\Temp\ipykernel_20144\1546052265.py:23: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_3 = pd.read_sql_query(query, connection)


,cnt
0,168


In [25]:
# Задание 6.4: Количество DS-вакансий с ключевым навыком Python
query = '''
SELECT COUNT(*) AS cnt
FROM public.vacancies
WHERE (
    name ILIKE '%data scientist%'
    OR name ILIKE '%data science%'
    OR name ILIKE '%исследователь данных%'
    OR name ~* '\\yml\\y'
    OR name ILIKE '%machine learning%'
    OR name ILIKE '%машинн%обучен%'
)
AND NOT (
    name ILIKE '%junior%'
    OR experience = 'Нет опыта'
    OR employment = 'Стажировка'
)
AND key_skills ILIKE '%python%'
'''
df_6_4 = pd.read_sql_query(query, connection)
df_6_4

C:\Users\Lenovo\AppData\Local\Temp\ipykernel_20144\3628686056.py:20: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_4 = pd.read_sql_query(query, connection)


,cnt
0,308


In [26]:
# Задание 6.5: Среднее количество ключевых навыков в DS-вакансиях
query = '''
SELECT ROUND(AVG(array_length(string_to_array(key_skills, E'\\t'), 1))::numeric, 2) AS avg_skills
FROM public.vacancies
WHERE (
    name ILIKE '%data scientist%'
    OR name ILIKE '%data science%'
    OR name ILIKE '%исследователь данных%'
    OR name ~* '\\yml\\y'
    OR name ILIKE '%machine learning%'
    OR name ILIKE '%машинн%обучен%'
)
AND NOT (
    name ILIKE '%junior%'
    OR experience = 'Нет опыта'
    OR employment = 'Стажировка'
)
'''
df_6_5 = pd.read_sql_query(query, connection)
df_6_5

C:\Users\Lenovo\AppData\Local\Temp\ipykernel_20144\3015283989.py:19: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_5 = pd.read_sql_query(query, connection)


,avg_skills
0,6.26


При выполнении заданий 6.3–6.5 обнаружено, что поле key_skills хранит навыки через символ табуляции (\t), а часть значений объединяет несколько технологий в одной строке (например, 'MS SQL/PostgreSQL', 'СУБД PostgreSQL', 'IPython', 'CPython'). Это создаёт неоднозначность при подсчёте: простой поиск подстроки (ILIKE '%SQL%', ILIKE '%python%') даёт результаты 168 и 308/310 соответственно, но такой подход рискует засчитывать смежные, но иные технологии (MySQL, IPython, CPython) как совпадения. Попытка точного посимвольного сравнения после разбиения по \t (unnest(string_to_array(...))) даёт более консервативные числа — 165 (SQL/postgres) и 305 (Python), а также 6.26 для среднего количества навыков (6.5) — но ни один из вариантов (в обоих направлениях трактовки) не был принят платформой. Контрольные проверки (совпадение промежуточных сумм, воспроизводимость фильтра DS/non-Junior на заданиях 6.2 и 6.6) подтверждают корректность базовой логики фильтрации; расхождение, по всей видимости, связано либо с иной трактовкой критерия "указан навык", либо с временной технической проблемой на платформе. Буду признателен за уточнение точного критерия при код-ревью.